### PR TITLE
Update experiment files

### PR DESCRIPTION
**Pull request by DeNeutoy**
@schmmd was having problems running our examples because they contained legacy configurations. This corrects those and adds some tests to check that all example experiments can be correctly loaded into a model without raising errors.

### CODE REVIEW

File: example_experiments/reading_comprehension/gareader_who_did_what.json [hunk 1]
```diff
@@ -1,11 +1,15 @@
 {
     "embeddings": {
       "words": {
-        "pretrained_file": "/efs/data/dlfa/glove/glove.6B.100d.txt.gz",
+        "pretrained_file": "/net/efs/data/dlfa/glove/glove.6B.100d.txt.gz",
         "fine_tune": false,
-        "project": false
+        "project": false,
+          "droppout": 0.0
```
Comment (matt-gardner) [inline example_experiments/reading_comprehension/gareader_who_did_what.json]: Spacing is off here.

File: example_experiments/reading_comprehension/gareader_who_did_what.json [hunk 2]
```diff
@@ -1,11 +1,15 @@
...
       },
-      "characters": {"dimension": 16}
+      "characters": {
+          "dimension": 16,
```
Comment (matt-gardner) [inline example_experiments/reading_comprehension/gareader_who_did_what.json]: Two space indents, I think.

Honestly, I don't much care which it is, but I noticed it because it's not consistent in the file.  I just checked other files, and the spacing isn't consistent across files.  It's really not worth standardizing all of them, but at least within a file, the indentation should be consistent.

File: example_experiments/entailment/snli_decomposable_attention.json [hunk 3]
```diff
@@ -21,12 +21,14 @@
...
     "batch_size": 60,
-    "pretrained_embeddings_file": "/net/efs/aristo/dlfa/glove/glove.840B.300d.txt.gz",
-    "project_embeddings": true,
-    "fine_tune_embeddings": false,
-    "embedding_dropout": 0.2,
     "patience": 3,
-    "embeddings": {"words": {"dimension": 200}},
+    "embeddings": {
+      "words": {"dimension": 200},
+      "pretrained_file": "/net/efs/aristo/dlfa/glove/glove.840B.300d.txt.gz",
```
Comment (matt-gardner) [inline example_experiments/entailment/snli_decomposable_attention.json]: This is wrong.  If it's supposed to be passing tests, the test isn't working.

File: example_experiments/multiple_choice_qa/simple_experiment.json [hunk 4]
```diff
@@ -6,9 +6,12 @@
             "type": "bow"
         }
     },
-    "pretrained_embeddings_file": "/efs/data/dlfa/glove/glove.840B.300d.txt.gz",
-    "project_embeddings": false,
-    "fine_tune_embeddings": false,
+    "embeddings":{
```
Comment (matt-gardner) [inline example_experiments/multiple_choice_qa/simple_experiment.json]: Also wrong - it's missing a top-level "words" key.

File: example_experiments/sequence_tagging/simple_tagger.json [hunk 5]
```diff
@@ -19,19 +19,22 @@
...
-    "embedding_dropout": 0.2,
     "data_generator": {
       "dynamic_padding": true
     },
     "patience": 3,
-    "embeddings": {"words": {"dimension": 100}, "characters": {"dimension": 8}},
+    "embeddings": {
+      "words": {"dimension": 100},
+      "characters": {"dimension": 8},
+      "pretrained_file": "/net/efs/data/dlfa/glove/glove.6B.100d.txt.gz",
```
Comment (matt-gardner) [inline example_experiments/sequence_tagging/simple_tagger.json]: These four lines go inside "words".

File: tests/example_experiments_test.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+    def check_experiment_type_can_be_parsed(self, directory):
+        for param_file in os.listdir(directory):
+            param_path = os.path.join(directory, param_file)
+            param_dict = pyhocon.ConfigFactory.parse_file(param_path)
+            params = Params(replace_none(param_dict))
+            model_class = concrete_models[params.pop("model_class")]
+            # Tests will try to create root directories as we have /net/efs paths,
+            # so we just remove the serialisation aspect here.
+            params["model_serialization_prefix"] = None
+            self.get_model(model_class, params)
```
Comment (matt-gardner) [inline tests/example_experiments_test.py]: Maybe you if you also built the model here, you'd get more parameters read, and it would find the errors?  You'd have to do some mangling somewhere, though, because `_build_model()` will try to read the embedding file.  Probably you could check for `/net/efs/`, and replace it with `self.TEST_DIR`, then make sure that a toy embedding file has been written to the right place.

Comment (DeNeutoy) [inline tests/example_experiments_test.py]: Yeah good point. I'd do this now.

Comment (matt-gardner) [inline tests/example_experiments_test.py]: Oh, actually this won't work.  You need to have some things set from the data file before the model can be built.  You'd need to mangle all of the train and validation files, too, and make sure that the correct files were written to `self.TRAIN_FILE`, etc...

This might be more work than it's worth.  It _would_ be really nice to have tests that our example files pass, though.  A better option might be to have some fixtures checked in to the repo for the tests, and have the examples point to those toy fixture files.

Comment (matt-gardner) [inline tests/example_experiments_test.py]: Then you just make sure that the example files train correctly with the fixture files.

**Comment by schmmd** [comment]
I can confirm that I'm getting much farther with the experiments on both my Kubernetes node (https://github.com/allenai/deep_qa/pull/379) and my local machine.

File: tests/example_experiments_test.py (hunk 6 above)
Comment (DeNeutoy) [inline tests/example_experiments_test.py]: It's kind of useful to have the example files pointing to actual stuff on efs though ...
One other way we could do this is set up a regression test docker image, which runs just this test actually on efs say once a week, with the result emailed to deep_qa@allenai.org or something. That would have the added benefit of checking that the paths are correct.  Maybe that's overkill though.

Comment (matt-gardner) [inline tests/example_experiments_test.py]: Or we could have actual paths in comments above the fixture path?

Comment (DeNeutoy) [inline tests/example_experiments_test.py]: yeah that's easier.

File: deep_qa/training/text_trainer.py [hunk 7]
```diff
@@ -658,12 +659,12 @@ def __get_new_embedding(self, name: str, vocab_name: str='words'):
...
                 embedding_layer = TimeDistributedEmbedding(
                         input_dim=self.data_indexer.get_vocab_size(vocab_name),
-                        output_dim=embedding_params.pop('dimension'),
+                        output_dim=embedding_dim,
                         mask_zero=True,  # this handles padding correctly
                         name=name + '_embedding')
             projection_layer = None
             if embedding_params.pop('project', False):
-                projection_layer = TimeDistributed(Dense(units=embedding_params.pop('dimension'),),
+                projection_layer = TimeDistributed(Dense(units=embedding_dim,),
```
Comment (matt-gardner) [inline deep_qa/training/text_trainer.py]: I'm confused as to why this was a bug.  Do you really have a case where you project non-pretrained embeddings?  We might actually want to fail here, because you probably don't want to do this.

Comment (DeNeutoy) [inline deep_qa/training/text_trainer.py]: Sorry, my head is not screwed on properly today, obviously we don't want this to happen. I'll make it raise an error as it was just throwing a param.pop error as the key was being accessed twice.

**Comment by schmmd** [comment]
@DeNeutoy I did have an exception running one of the experiments even with your pull request.

```
# python scripts/run_model.py example_experiments/multiple_choice_qa/simple_experiment.json
2017-06-01 22:01:12,712 - PARAM - deep_qa.common.params - random_seed = 13370
2017-06-01 22:01:12,712 - PARAM - deep_qa.common.params - numpy_seed = 1337
Using TensorFlow backend.
2017-06-01 22:01:15,092 - INFO - deep_qa.common.checks - Keras version: 2.0.4
2017-06-01 22:01:15,093 - INFO - deep_qa.common.checks - Tensorflow version: 1.1.0
2017-06-01 22:01:15,596 - PARAM - deep_qa.common.params - processor = {}
2017-06-01 22:01:15,597 - PARAM - deep_qa.common.params - processor.word_splitter = simple
2017-06-01 22:01:15,597 - PARAM - deep_qa.common.params - processor.word_filter = pass_through
2017-06-01 22:01:15,597 - PARAM - deep_qa.common.params - processor.word_stemmer = pass_through
2017-06-01 22:01:15,672 - PARAM - deep_qa.common.params - log_device_placement = False
2017-06-01 22:01:15.673497: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.1 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-01 22:01:15.673538: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use SSE4.2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-01 22:01:15.673564: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX instructions, but these are available on your machine and could speed up CPU computations.
2017-06-01 22:01:15.673587: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use AVX2 instructions, but these are available on your machine and could speed up CPU computations.
2017-06-01 22:01:15.673609: W tensorflow/core/platform/cpu_feature_guard.cc:45] The TensorFlow library wasn't compiled to use FMA instructions, but these are available on your machine and could speed up CPU computations.
2017-06-01 22:01:15.837388: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:901] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2017-06-01 22:01:15.837970: I tensorflow/core/common_runtime/gpu/gpu_device.cc:887] Found device 0 with properties: 
name: Tesla K80
major: 3 minor: 7 memoryClockRate (GHz) 0.8235
pciBusID 0000:00:1e.0
Total memory: 11.17GiB
Free memory: 11.11GiB
2017-06-01 22:01:15.838024: I tensorflow/core/common_runtime/gpu/gpu_device.cc:908] DMA: 0 
2017-06-01 22:01:15.838055: I tensorflow/core/common_runtime/gpu/gpu_device.cc:918] 0:   Y 
2017-06-01 22:01:15.838086: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977] Creating TensorFlow device (/gpu:0) -> (device: 0, name: Tesla K80, pci bus id: 0000:00:1e.0)
2017-06-01 22:01:15,921 - PARAM - deep_qa.common.params - model_class = MultipleTrueFalseMemoryNetwork
2017-06-01 22:01:15,921 - PARAM - deep_qa.common.params - num_options = None
2017-06-01 22:01:15,921 - PARAM - deep_qa.common.params - num_memory_layers = 1
2017-06-01 22:01:15,922 - PARAM - deep_qa.common.params - knowledge_encoder = {}
2017-06-01 22:01:15,922 - PARAM - deep_qa.common.params - knowledge_selector = ConfigTree([('type', 'dot_product')])
2017-06-01 22:01:15,922 - PARAM - deep_qa.common.params - knowledge_combiner = {}
2017-06-01 22:01:15,922 - PARAM - deep_qa.common.params - memory_updater = ConfigTree([('type', 'sum')])
2017-06-01 22:01:15,922 - PARAM - deep_qa.common.params - entailment_input_combiner = ConfigTree([('type', 'memory_only')])
2017-06-01 22:01:15,923 - PARAM - deep_qa.common.params - entailment_model = {}
2017-06-01 22:01:15,923 - PARAM - deep_qa.common.params - recurrence_mode = {}
2017-06-01 22:01:15,923 - PARAM - deep_qa.common.params - max_knowledge_length = None
2017-06-01 22:01:15,923 - PARAM - deep_qa.common.params - embeddings = ConfigTree([('words', ConfigTree([('pretrained_file', '/net/efs/data/dlfa/glove/glove.840B.300d.txt.gz'), ('project', False), ('fine_tune', False)]))])
2017-06-01 22:01:15,923 - PARAM - deep_qa.common.params - data_generator = None
2017-06-01 22:01:15,924 - PARAM - deep_qa.common.params - num_sentence_words = 125
2017-06-01 22:01:15,924 - PARAM - deep_qa.common.params - num_word_characters = None
2017-06-01 22:01:15,924 - PARAM - deep_qa.common.params - tokenizer = {}
2017-06-01 22:01:15,924 - PARAM - deep_qa.common.params - tokenizer.type = words
2017-06-01 22:01:15,924 - PARAM - deep_qa.common.params - tokenizer.processor = {}
2017-06-01 22:01:15,924 - PARAM - deep_qa.common.params - tokenizer.processor.word_splitter = simple
2017-06-01 22:01:15,925 - PARAM - deep_qa.common.params - tokenizer.processor.word_filter = pass_through
2017-06-01 22:01:15,925 - PARAM - deep_qa.common.params - tokenizer.processor.word_stemmer = pass_through
2017-06-01 22:01:15,925 - PARAM - deep_qa.common.params - encoder = ConfigTree([('default', ConfigTree([('type', 'bow')]))])
2017-06-01 22:01:15,925 - PARAM - deep_qa.common.params - encoder_fallback_behavior = crash
2017-06-01 22:01:15,925 - PARAM - deep_qa.common.params - seq2seq_encoder = {'default': {'encoder_params': {}, 'wrapper_params': {}}}
2017-06-01 22:01:15,926 - PARAM - deep_qa.common.params - seq2seq_encoder_fallback_behavior = crash
2017-06-01 22:01:15,926 - PARAM - deep_qa.common.params - train_files = ['data/science/omnibus_questions/processed_questions.tsv', 'data/science/omnibus_questions/processed_questions_background.tsv']
2017-06-01 22:01:15,926 - PARAM - deep_qa.common.params - validation_files = None
2017-06-01 22:01:15,926 - PARAM - deep_qa.common.params - test_files = None
2017-06-01 22:01:15,926 - PARAM - deep_qa.common.params - max_training_instances = None
2017-06-01 22:01:15,926 - PARAM - deep_qa.common.params - max_validation_instances = None
2017-06-01 22:01:15,927 - PARAM - deep_qa.common.params - max_test_instances = None
2017-06-01 22:01:15,927 - PARAM - deep_qa.common.params - train_steps_per_epoch = None
2017-06-01 22:01:15,927 - PARAM - deep_qa.common.params - train_steps_per_epoch = None
2017-06-01 22:01:15,927 - PARAM - deep_qa.common.params - train_steps_per_epoch = None
2017-06-01 22:01:15,927 - PARAM - deep_qa.common.params - save_models = True
2017-06-01 22:01:15,927 - PARAM - deep_qa.common.params - model_serialization_prefix = models/example
2017-06-01 22:01:15,928 - PARAM - deep_qa.common.params - num_gpus = 1
2017-06-01 22:01:15,928 - PARAM - deep_qa.common.params - validation_split = 0.1
2017-06-01 22:01:15,928 - PARAM - deep_qa.common.params - batch_size = 32
2017-06-01 22:01:15,928 - PARAM - deep_qa.common.params - num_epochs = 20
2017-06-01 22:01:15,928 - PARAM - deep_qa.common.params - optimizer = adam
2017-06-01 22:01:15,928 - PARAM - deep_qa.common.params - gradient_clipping = {'value': 10, 'type': 'clip_by_norm'}
2017-06-01 22:01:15,928 - PARAM - deep_qa.common.params - loss = categorical_crossentropy
2017-06-01 22:01:15,929 - PARAM - deep_qa.common.params - metrics = ['accuracy']
2017-06-01 22:01:15,929 - PARAM - deep_qa.common.params - validation_metric = val_acc
2017-06-01 22:01:15,929 - PARAM - deep_qa.common.params - patience = 5
2017-06-01 22:01:15,929 - PARAM - deep_qa.common.params - fit_kwargs = {}
2017-06-01 22:01:15,929 - PARAM - deep_qa.common.params - tensorboard_log = None
2017-06-01 22:01:15,929 - PARAM - deep_qa.common.params - tensorboard_frequency = 0
2017-06-01 22:01:15,930 - PARAM - deep_qa.common.params - debug = {}
2017-06-01 22:01:15,930 - PARAM - deep_qa.common.params - show_summary_with_masking_info = True
2017-06-01 22:01:15,930 - INFO - deep_qa.run - Training model
2017-06-01 22:01:15,930 - INFO - deep_qa.training.trainer - Running training (MemoryNetwork)
Traceback (most recent call last):
  File "scripts/run_model.py", line 35, in <module>
    main()
  File "scripts/run_model.py", line 16, in main
    run_model(sys.argv[1])
  File "scripts/../deep_qa/run.py", line 109, in run_model
    model.train()
  File "scripts/../deep_qa/training/trainer.py", line 278, in train
    self.training_dataset = self.load_dataset_from_files(self.train_files)
  File "scripts/../deep_qa/models/multiple_choice_qa/multiple_true_false_memory_network.py", line 107, in load_dataset_from_files
    dataset = super(MultipleTrueFalseMemoryNetwork, self).load_dataset_from_files(files)
  File "scripts/../deep_qa/models/memory_networks/memory_network.py", line 141, in load_dataset_from_files
    dataset = super(MemoryNetwork, self).load_dataset_from_files(files)
  File "scripts/../deep_qa/training/text_trainer.py", line 175, in load_dataset_from_files
    return TextDataset.read_from_file(files[0], self._instance_type())
  File "scripts/../deep_qa/data/dataset.py", line 75, in read_from_file
    with codecs.open(filename, 'r', 'utf-8') as input_file:
  File "/opt/conda/envs/runenv/lib/python3.5/codecs.py", line 895, in open
    file = builtins.open(filename, mode, buffering)
FileNotFoundError: [Errno 2] No such file or directory: 'data/science/omnibus_questions/processed_questions.tsv'
Exception ignored in: <bound method BaseSession.__del__ of <tensorflow.python.client.session.Session object at 0x7f9d59a0a7f0>>
Traceback (most recent call last):
  File "/opt/conda/envs/runenv/lib/python3.5/site-packages/tensorflow/python/client/session.py", line 587, in __del__
AttributeError: 'NoneType' object has no attribute 'TF_NewStatus'
```

**Comment by DeNeutoy** [comment]
Ok thanks, that data is something I think most people had locally originally, as it was small scale testing data - I didn't update the path the first time through this PR because I wasn't sure which dataset it actually corresponded to - I will hunt it down now.

**Comment by DeNeutoy** [comment]
@matt-gardner This is ready for a second look - I haven't touched the default Memory Network embedding/encoder issue we discussed yesterday.

Basically i've moved all semi-duplicated experiments out of `example_experiments` into `contrib/example_experiments` and all of the ones remaining have tests. The one thing this doesn't test for is that the `/efs` filepaths are correct, but there's not much we can do about that.

**Comment by DeNeutoy** [comment]
Also these tests are quite slow - testing all of them takes 2min.

File: tests/example_experiments_test.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+        super(TestExampleExperiments, self).setUp()
+        self.example_experiments_dir = "./example_experiments"
+        self.entailment_dir = self.example_experiments_dir + "/entailment/"
+        self.memory_networks_dir = self.example_experiments_dir + "/memory_networks/"
+        self.multiple_choice_qa_dir = self.example_experiments_dir + "/multiple_choice_qa/"
+        self.reading_comprehension_dir = self.example_experiments_dir + "/reading_comprehension/"
+        self.sentence_selection_dir = self.example_experiments_dir + "/sentence_selection/"
+        self.sequence_tagging_dir = self.example_experiments_dir + "/sequence_tagging/"
+
+    def test_entailment_examples_can_be_parsed(self):
```
Comment (matt-gardner) [inline tests/example_experiments_test.py]: I would change all of these `can_be_parsed` names to `can_train`, as that's what we're actually doing now.

File: tests/example_experiments_test.py [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        params["model_serialization_prefix"] = None
+        if len(params["train_files"]) > 1:
+            params["train_files"] = [self.TRAIN_FILE, self.TRAIN_BACKGROUND]
+            params["validation_files"] = [self.VALIDATION_FILE, self.VALIDATION_BACKGROUND]
+        else:
+            params["train_files"] = [self.TRAIN_FILE]
+            params["validation_files"] = [self.TRAIN_FILE]
+        params["num_epochs"] = 1
+        try:
+            params["embeddings"]["words"]["project"] = False
```
Comment (matt-gardner) [inline tests/example_experiments_test.py]: How does this work for experiments that used pre-trained embeddings?  It seems like this should be crashing for some of the experiments.  The simple tagger, e.g., should be trying to read the pre-trained embedding file, and it should be crashing because it's not present.  What's going on?

File: tests/example_experiments_test.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+from deep_qa.common.params import Params, replace_none
+from deep_qa.models import concrete_models
+
+from .common.test_case import DeepQaTestCase
+
+
+class TestExampleExperiments(DeepQaTestCase):
+    def setUp(self):
+        super(TestExampleExperiments, self).setUp()
+        self.example_experiments_dir = "./example_experiments"
```
Comment (matt-gardner) [inline tests/example_experiments_test.py]: Not sure this setup is buying us much at this point, as each experiment file is tested individually now.  It'd probably be simpler to just give the full path in a test directly.  Then it's easier to think about adding new ones, and checking if one is there.  Fewer places to modify or look at when you're adding a new test.

File: deep_qa/common/util.py [hunk 11]
```diff
@@ -30,3 +30,42 @@ def add_noise_to_dict_values(dictionary: Dict[Any, float], noise_param: float) -
         noise = random.uniform(-noise_value, noise_value)
         new_dict[key] = value + noise
     return new_dict
+
+
+def clean_layer_name(input_name: str,
+                     keep_right_of_last_backslash: bool=None,
```
Comment (matt-gardner) [inline deep_qa/common/util.py]: This name is confusing to me.  Seems like you want a parameter named "split on backslash", or "remove nested scopes", or something, with only one behavior - this is never called with `True`, only with `False` and `None`, so we can simplify this and name it something more appropriate.

File: deep_qa/common/util.py [hunk 12]
```diff
@@ -30,3 +30,42 @@ def add_noise_to_dict_values(dictionary: Dict[Any, float], noise_param: float) -
...
+    input_name: str, required
+        A Keras layer name.
+    keep_right_of_last_backslash: bool or None, optional, (default = None)
+        If we split on the last backslash, which side should we keep? This is
+        useful in the case that you want to preserve a particular aspect of a scope,
+        but don't care about the other part.
+    strip_numerics_after_underscores: bool, optional, (default = True)
+        If there are numerical values after an underscore at the end of the layer name,
+        this flag specifies whether or not to remove them.
+
```
Comment (matt-gardner) [inline deep_qa/common/util.py]: A bunch of unnecessary whitespace here.

File: deep_qa/data/tokenizers/word_and_character_tokenizer.py [hunk 13]
```diff
@@ -116,13 +117,7 @@ def embed_input(self,
         # input layer.  Except sometimes Keras adds funny things to the ends of the input layer, so
         # we'll strip those off.
         input_name = input_layer.name
```
Comment (matt-gardner) [inline deep_qa/data/tokenizers/word_and_character_tokenizer.py]: We probably don't need this intermediate variable anymore.

File: deep_qa/training/text_trainer.py [hunk 14]
```diff
@@ -624,9 +625,12 @@ def __get_embedded_input(self,
 
         embedding_layer, projection_layer, dropout = self.embedding_layers[embedding_name]
         embedded_input = embedding_layer(input_layer)
+
+        layer_name = clean_layer_name(input_layer.name, strip_numerics_after_underscores=False)
         if projection_layer is not None:
-            for _ in range(2, K.ndim(input_layer)):  # 2 here to account for batch_size.
-                projection_layer = TimeDistributed(projection_layer, name="timedist_" + projection_layer.name)
+            for i in range(1, K.ndim(input_layer)):  # 2 here to account for batch_size.
```
Comment (matt-gardner) [inline deep_qa/training/text_trainer.py]: Comment needs to be updated here.

File: deep_qa/training/text_trainer.py [hunk 15]
```diff
@@ -647,31 +651,47 @@ def __get_new_embedding(self, name: str, vocab_name: str='words'):
...
+            projection_layer = None
             if pretrained_file:
                 embedding_layer = PretrainedEmbeddings.get_embedding_layer(
                         pretrained_file,
                         self.data_indexer,
                         embedding_params.pop('fine_tune', False),
                         name=name + '_embedding')
+
+                if embedding_params.pop('project', False):
+                    projection_layer = Dense(units=embedding_params.pop('dimension'), name=name + "_projection")
```
Comment (matt-gardner) [inline deep_qa/training/text_trainer.py]: Maybe add a comment for why there's no `TimeDistributed` here - we'll handle it later, so that we can more easily reuse the embeddings on different input lengths.

File: tests/data/embeddings_test.py [hunk 16]
```diff
@@ -66,3 +68,38 @@ def test_get_embedding_layer_initializes_unseen_words_randomly_not_zero(self):
...
+                                    "project": True,
+                                    "fine_tune": True,
+                                    "dropout": 0.2
+                            }
+                    }
+            }
+            model = self.get_model(ClassificationModel, args)
+            model.train()
+
+    def test_projection_vs_pretrained_dim_mismatch_raises_error(self):
```
Comment (matt-gardner) [inline tests/data/embeddings_test.py]: What's the "vs" mean here?

File: deep_qa/common/util.py (hunk 11 above)
Comment (DeNeutoy) [inline deep_qa/common/util.py]: If it's None, we don't split at all though - if it's true, we take the right hand side after splitting, if false we take the left hand side, but I agree that it's confusing.

Comment (matt-gardner) [inline deep_qa/common/util.py]: Right, I was just saying that taking the right hand side after splitting is currently unused, so we might as well not add the complexity if it's not needed.

Comment (DeNeutoy) [inline deep_qa/common/util.py]: Ah it should have been, thanks for catching!

File: tests/data/embeddings_test.py (hunk 16 above)
Comment (DeNeutoy) [inline tests/data/embeddings_test.py]: "versus", changed it to something better.
